### PR TITLE
KUZ-541 Allow plugins to register an event on multiple functions

### PR DESCRIPTION
**Pull request by scottinet**
This pull request allows a plugin to be plugged on an event with different functions.

The `hooks` and `pipes` properties now accept either a string or an array of strings:
```js
{
  hooks: {
    'foo:bar': 'someMethod',
    'bar:foo': ['some', 'other', 'methods']
  }
}
```

Unit tests on worker plugins and on the `pluginsManager.run()` method have also been rewritten to use Sinon.js instead of the handmade mocks . Well... except for that humongous PM2 mock, I didn't have the courage to refactor this one.

### CODE REVIEW

File: lib/api/core/plugins/pluginsManager.js [hunk 1]
```diff
@@ -313,60 +313,78 @@ function pm2Init (workers, pluginsWorker, isDummy) {
 }
 
 function initPipes (pipes, plugin, pipeWarnTime, pipeTimeout) {
-  if (plugin.config && plugin.config.pipeWarnTime !== undefined) {
-    pipeWarnTime = plugin.config.pipeWarnTime;
-  }
-  if (plugin.config && plugin.config.pipeTimeout !== undefined) {
-    pipeTimeout = plugin.config.pipeTimeout;
-  }
+  var register = (event, fn) => {
```
Comment (AnthonySendra) [inline lib/api/core/plugins/pluginsManager.js]: Generally, I'm not really comfortable with function creation inside other functions. It's just a personal point of view, but can we create instead a private function inside the same file? Something like:
`registerPipes`
`registerHooks`

I find it difficult to read and to understand. We can discuss if you're not agree.

**Comment by ballinette** [comment]
+1

File: lib/api/core/plugins/pluginsManager.js (hunk 1 above)
Comment (j33f) [inline lib/api/core/plugins/pluginsManager.js]: I agree

Comment (scottinet) [inline lib/api/core/plugins/pluginsManager.js]: I'm fine with this too. Inc another commit :-)

**Comment by AnthonySendra** [comment]
+1

**Comment by dbengsch** [comment]
lib/api/core/plugins/workerReady.js and its tests should also be adapted in order to call more than one function at a time (line 40 to 43)

**Pull request merged**